File: Balanced_string.cpp
#include<iostream>
#include<stack>
using namespace std;
bool balanced_string(string str){
    stack<char> st;
    for(int i=0;i<str.length();i++){
        if(str[i] == '[' || str[i] == '(' || str[i] == '{')
            st.push(str[i]);
        else{
            if(!st.empty()){
                if(st.top() == '(' && str[i] == ')')
                    st.pop();
                else if(st.top() == '[' && str[i] == ']')
                    st.pop();
                else if(st.top() == '{' && str[i] == '}')
                    st.pop();
                else
                    return false;
            }
            else
                return false;
        }
    }
    if(st.empty())
        return true;
    return false;
}
int main(){
    string str;
    cin>>str;
    if(balanced_string(str))
        cout<<"YES";
    else
        cout<<"NO";
    return 0;
}

File: Balanced_string.py
# -*- coding: utf-8 -*-
"""
Created on Sat Oct 31 13:34:21 2020

@author: Shubham
"""
def balanced_string(string):
    stack = []
    for i in range(len(string)):
        if string[i] == '[' or string[i] == '{' or string[i] == '(' :
            stack.append(string[i])
        else:
            if len(stack) != 0:
                if stack[-1] == '(' and string[i] == ')' :
                    stack.pop()
                elif stack[-1] == '[' and string[i] == ']' :
                    stack.pop()
                elif stack[-1] == '{' and string[i] == '}' :
                    stack.pop()
                else :
                    return False
            else :
                return False
    if len(stack) == 0:
        return True
    return False
            
        
        
string = input()
if balanced_string(string):
    print("YES")
else:
    print("NO")